### ISSUE TITLE
Undefined symbol with version 0.5.3

### ISSUE THREAD

**Issue by Crassirostris** [GoogleCloudPlatform/fluent-plugin-google-cloud#106]
When installing `fluent-plugin-google-cloud` with version `0.5.3` on the most recent td-agent, using Ruby 2.1, the following error shows up during startup and fluentd exits

```
/opt/td-agent/embedded/bin/ruby: symbol lookup error: /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/google-protobuf-3.0.2-x86_64-linux/lib/google/protobuf_c.so: undefined symbol: rb_data_typed_object_wrap
```

Minimal configuration to reproduce

```
<match **>
  type google_cloud
</match>
```

From https://abi-laboratory.pro/tracker/compat_report/ruby/2.2.4/2.3.0/240f7/abi_compat_report.html it seems that protobuf_c.so requires Ruby >= 2.3.0.

@igorpeshansky @piosz

**Comment by dhrupadb** [comment]
We found this is likely an issue with the google-protobuf gem being incompatible on older systems. We build a version ourselves which seems to solve the issue. I've attached the gem here. If you could confirm that replacing the packaged google protobuf gem solves this issue then we will go ahead and package it up in our builds until the protobuf issue is resolved. 

http://testrepo.stackdriver.com/gem/google-protobuf-3.0.2.gem

**Comment by Crassirostris** [comment]
@dhrupadb OK, thanks a lot! I will try this out soon

**Comment by ldelossa** [comment]
I'm suffering from this issue also. 

```
/opt/td-agent/embedded/bin/ruby: symbol lookup error: /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/google-protobuf-3.0.2-x86_64-linux/lib/google/protobuf_c.so: undefined symbol: rb_data_typed_object_wrap
```

Looks like it is referring to 3.0.2

**Comment by igorpeshansky** [comment]
@ldelossa, could you please try the `google-protobuf-3.0.2` gem linked in the comment [above](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/issues/106#issuecomment-273552783)?

**Comment by ldelossa** [comment]
How can I install this into td-agent's gem distribution?

**Comment by ldelossa** [comment]
```
root@player-sentry-prod-web-1-0-v82z:/tmp# td-agent-gem install google-protobuf-3.0.2.gem
Done installing documentation for  after 0 seconds
root@player-sentry-prod-web-1-0-v82z:/tmp# td-agent --config /etc/td-agent/td-agent.conf
2017-01-19 17:25:46 +0000 [info]: reading config file path="/etc/td-agent/td-agent.conf"
2017-01-19 17:25:46 +0000 [info]: starting fluentd-0.12.31
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-mixin-config-placeholders' version '0.4.0'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-mixin-plaintextformatter' version '0.2.6'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-google-cloud' version '0.5.3'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-kafka' version '0.4.1'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-mongo' version '0.7.16'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '1.5.5'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-s3' version '0.8.0'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-scribe' version '0.10.14'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-td' version '0.10.29'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-td-monitoring' version '0.2.2'
2017-01-19 17:25:46 +0000 [info]: gem 'fluent-plugin-webhdfs' version '0.4.2'
2017-01-19 17:25:46 +0000 [info]: gem 'fluentd' version '0.12.31'
2017-01-19 17:25:46 +0000 [info]: adding match pattern="td.sentry.accesslog" type="google_cloud"
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:3: warning: already initialized constant JSON::VERSION
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:3: warning: previous definition of VERSION was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:4: warning: already initialized constant JSON::VERSION_ARRAY
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:4: warning: previous definition of VERSION_ARRAY was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:5: warning: already initialized constant JSON::VERSION_MAJOR
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:5: warning: previous definition of VERSION_MAJOR was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:6: warning: already initialized constant JSON::VERSION_MINOR
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:6: warning: previous definition of VERSION_MINOR was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:7: warning: already initialized constant JSON::VERSION_BUILD
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:7: warning: previous definition of VERSION_BUILD was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:99: warning: already initialized constant JSON::NaN
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:99: warning: previous definition of NaN was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:101: warning: already initialized constant JSON::Infinity
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:101: warning: previous definition of Infinity was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:103: warning: already initialized constant JSON::MinusInfinity
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:103: warning: previous definition of MinusInfinity was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:128: warning: already initialized constant JSON::UnparserError
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:128: warning: previous definition of UnparserError was here
/opt/td-agent/embedded/bin/ruby: symbol lookup error: /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/google-protobuf-3.0.2-x86_64-linux/lib/google/protobuf_c.so: undefined symbol: rb_data_typed_object_wrap
2017-01-19 17:25:46 +0000 [info]: process finished code=32512
2017-01-19 17:25:46 +0000 [warn]: process died within 1 second. exit.
```

Think we have the same issue.

**Comment by dhrupadb** [comment]
You're going to have to uninstall the existing google-protobuf gem first. Because they're the same version I don't think you're actually picking up the second one without uninstalling the first. google-protobuf-3.0.2-x86_64 is the buggy gem you need to remove.

**Comment by mats116** [comment]
Is this issue resolved?

The same phenomenon is happening also in `fluent-plugin-google-cloud (0.6.0)` with `google-protobuf (3.2.0.2 x86_64-linux)` .

```shell
# service td-agent restart

Restarting td-agent: /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:3: warning: already initialized constant JSON::VERSION
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:3: warning: previous definition of VERSION was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:4: warning: already initialized constant JSON::VERSION_ARRAY
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:4: warning: previous definition of VERSION_ARRAY was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:5: warning: already initialized constant JSON::VERSION_MAJOR
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:5: warning: previous definition of VERSION_MAJOR was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:6: warning: already initialized constant JSON::VERSION_MINOR
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:6: warning: previous definition of VERSION_MINOR was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:7: warning: already initialized constant JSON::VERSION_BUILD
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:7: warning: previous definition of VERSION_BUILD was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:99: warning: already initialized constant JSON::NaN
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:99: warning: previous definition of NaN was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:101: warning: already initialized constant JSON::Infinity
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:101: warning: previous definition of Infinity was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:103: warning: already initialized constant JSON::MinusInfinity
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:103: warning: previous definition of MinusInfinity was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:128: warning: already initialized constant JSON::UnparserError
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:128: warning: previous definition of UnparserError was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:3: warning: already initialized constant JSON::VERSION
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:3: warning: previous definition of VERSION was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:4: warning: already initialized constant JSON::VERSION_ARRAY
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:4: warning: previous definition of VERSION_ARRAY was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:5: warning: already initialized constant JSON::VERSION_MAJOR
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:5: warning: previous definition of VERSION_MAJOR was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:6: warning: already initialized constant JSON::VERSION_MINOR
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:6: warning: previous definition of VERSION_MINOR was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/version.rb:7: warning: already initialized constant JSON::VERSION_BUILD
/opt/td-agent/embedded/lib/ruby/2.1.0/json/version.rb:7: warning: previous definition of VERSION_BUILD was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:99: warning: already initialized constant JSON::NaN
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:99: warning: previous definition of NaN was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:101: warning: already initialized constant JSON::Infinity
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:101: warning: previous definition of Infinity was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:103: warning: already initialized constant JSON::MinusInfinity
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:103: warning: previous definition of MinusInfinity was here
/opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.1/lib/json/common.rb:128: warning: already initialized constant JSON::UnparserError
/opt/td-agent/embedded/lib/ruby/2.1.0/json/common.rb:128: warning: previous definition of UnparserError was here
td-agent                                                   [  OK  ]
```

```shell
# /opt/td-agent/embedded/bin/fluent-gem list |grep fluentd

fluentd (0.12.31)
```

**Comment by crassirostris** [comment]
@mats116 You have only warnings, IIUC, that's irritating, but WAI

Before we had an error and a failure to start:

```
/opt/td-agent/embedded/bin/ruby: symbol lookup error: /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/google-protobuf-3.0.2-x86_64-linux/lib/google/protobuf_c.so: undefined symbol: rb_data_typed_object_wrap
```

This particular problem should be gone, isn't it?

**Comment by dhrupadb** [comment]
@mats116 - the logspam is caused due to two conflicting json gems that are
pulled in and itself isn't a problem. We intend on cleaning this up. The
Google protobuf issue should be resolved in the 3.2.0.2 gem.
